File: Cpp_Solutions/solution4.cpp
#include <iostream>
#include <string>
#include <algorithm>

// User input
void reverseString(std::string& str) {
    std::reverse(str.begin(), str.end());
}
// End of user input

class Checker {
public:
    static void check(void (*func)(std::string&), std::string input, const std::string& expected) {
        func(input);
        if (input == expected) {
            std::cout << "Test Case Passed: Result = " << input 
                      << ", Expected = " << expected << std::endl;
        } else {
            std::cout << "Failed Test Case: Result = " << input 
                      << ", Expected = " << expected << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <string> <expected>" << std::endl;
        return 1;
    }

    std::string input = argv[1];
    std::string expected = argv[2];

    Checker::check(reverseString, input, expected);
    return 0;
} 

File: Solutions/Problem4/Solution4.java
public class Solution4 {
    // User input
    public static String reverseString(String str) {
        return new StringBuilder(str).reverse().toString();
    }
    // End of user input

    static class Checker {
        public static void check(String input, String expected) {
            String result = reverseString(input);
            if (result.equals(expected)) {
                System.out.printf("Test Case Passed: Result = %s, Expected = %s%n", 
                                result, expected);
            } else {
                System.out.printf("Failed Test Case: Result = %s, Expected = %s%n", 
                                result, expected);
            }
        }
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.printf("Usage: java Solution4 <string> <expected>%n");
            System.exit(1);
        }

        String input = args[0];
        String expected = args[1];

        Checker.check(input, expected);
    }
} 